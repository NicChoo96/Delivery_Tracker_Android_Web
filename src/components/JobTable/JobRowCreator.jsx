import React, { useEffect, useState, } from 'react';
import {
  TableCell,
  TableRow,
} from '@material-ui/core';
import { LazyJobRow } from '../../components';
import { immutableSort, fuzzySearch } from '../../utils';

/**
 * Sorts and Filters the jobs to display before rendering the rows for JobTable
 * @category UI Component
 * @subcategory JobTable
 * @class JobRowCreator
 * @param {Array<jobInformation>} jobData array of all jobs
 * @param {boolean} showCompleted render completed jobs
 * @param {boolean} showInCompleted render incompleted jobs
 * @param {boolean} descending render jobs in descending date time order
 * @param {string} searchTerm input search term to fuzzy search from contractor name, company and phone number
 * @return {ReactComponent}
 */
const JobRowCreator = ({ jobData, showCompleted, showInCompleted, descending, searchTerm }) => {
  const [sortedData, setSortedData] = useState([]);
  const [displayData, setDisplayData] = useState([]);
  const sortingOrder = descending ? 1 : -1;

  // put all completed task at the bottom
  // sort tasks ascending with dateTimeStart
  // sort tasks ascending with dateTimeEnd
  useEffect(() => {
    setSortedData(immutableSort(jobData, (a, b) => {

      // completed tasks always comes later
      // only one has dateTimeEnd
      if (a.dateTimeEnd > 0 && b.dateTimeEnd === 0)
        return 1;
      else if (a.dateTimeEnd === 0 && b.dateTimeEnd > 0)
        return -1;

      // sort if datetimeStart are not the same
      if (a.dateTimeStart !== b.dateTimeStart) {
        return (b.dateTimeStart - a.dateTimeStart) * sortingOrder;
      }

      // if datetimeStart are the same, sort by datetimeEnd
      // both WILL have datetimeEnd at this point
      if (a.dateTimeEnd && b.dateTimeEnd)
        return (b.dateTimeEnd - a.dateTimeEnd) * sortingOrder;

      return 1;
    }));
  }, [jobData, sortingOrder]);

  // setting filters
  useEffect(() => {
    setDisplayData(sortedData.filter((data) => {
      const searchTermFound = fuzzySearch(searchTerm, `${data.name} ${data.company} ${data.phoneNo}`);
      if (showCompleted && showInCompleted) return searchTermFound;
      else if (showCompleted) return (data.dateTimeEnd > 0) && searchTermFound;
      else if (showInCompleted) return (data.dateTimeEnd === 0) && searchTermFound;
      return false;
    }));
  }, [showCompleted, showInCompleted, sortedData, searchTerm]);

  return (
    (displayData.length > 0)
      ? (<>{displayData.map((job, index) => (<LazyJobRow
        jobID={job?.jobID ?? '-'}
        location={job.locations}
        dateTimeStart={job.dateTimeStart}
        dateTimeEnd={job.dateTimeEnd}
        company={job.company}
        phone={job.phoneNo}
        contractorName={job.name}
        status={!job.dateTimeEnd ? 'In Progress' : 'Complete'}
        key={index}
      />))}</>)
      : <TableRow key={69}>
        <TableCell align="center" colSpan={8}>
          <span className="app__no-data-message">No data available.</span>
        </TableCell>
      </TableRow>
  )
};

export default JobRowCreator;
