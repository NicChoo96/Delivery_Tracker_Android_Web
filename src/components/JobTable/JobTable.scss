.job-table__data-control-group {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  margin-bottom: 1em;
  margin-left: 1em;
  user-select: none;
}
