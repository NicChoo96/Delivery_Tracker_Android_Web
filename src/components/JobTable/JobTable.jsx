import React, { Suspense, useState, } from 'react';
import {
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  FormGroup,
  FormControlLabel,
  TextField,
  Checkbox,
  Switch,
} from '@material-ui/core';
import { Loader } from '../../components';
import JobRowCreator from './JobRowCreator';

import "./JobTable.scss";

/**
 * Sorts and Filters the jobs to display before rendering the rows for JobTable
 * @category UI Component
 * @subcategory JobTable
 * @class JobTable
 * @param {Array<jobData>} data array of all job data
 * @return {ReactComponent}
 */
const JobTable = ({ data = [] }) => {
  const [displayCheckboxCompleted, setDisplayCheckboxCompleted] = useState(true);
  const [displayCheckboxInCompleted, setDisplayCheckboxInCompleted] = useState(true);
  const [sortDescending, setSortDescending] = useState(true);
  const [searchTerm, setSearchTerm] = useState('');

  const changeSetterHandler = (stateSetter) => (event) => {
    stateSetter(event.target.checked);
  }

  const searchHandler = (event) => {
    setSearchTerm(event.target.value);
  }

  return (
    <>
      <div className="job-table__data-control-group">
        <FormGroup row>
          <FormControlLabel
            checked={displayCheckboxInCompleted}
            onChange={changeSetterHandler(setDisplayCheckboxInCompleted)}
            control={<Checkbox color="primary" size="small" name="Show Completed" />}
            label="In Progress"
          />
          <FormControlLabel
            checked={displayCheckboxCompleted}
            onChange={(changeSetterHandler(setDisplayCheckboxCompleted))}
            control={<Checkbox color="primary" size="small" name="Show Incompleted" />}
            label="Completed"
          />
          <FormControlLabel
            control={
              <Switch
                checked={sortDescending}
                name="Show Descending"
                onChange={(changeSetterHandler(setSortDescending))}
                color="primary"
              />}
            label="Date Time Descending"
          />
        </FormGroup>
        <FormGroup row>
          <FormControlLabel
            onChange={searchHandler}
            control={<TextField size="small" label="Search for Contractor" variant="outlined" />}
          />
        </FormGroup>
      </div>
      <TableContainer component={Paper} className="app-table">
        <Table aria-label="a dense table" size="small">
          <TableHead>
            <TableRow>
              <TableCell align="left" style={{ width: '8em' }} ><span className="app-table__header-cell">Job ID</span></TableCell>
              <TableCell align="left"><span className="app-table__header-cell">Location</span></TableCell>
              <TableCell align="left"><span className="app-table__header-cell">Date Time Started</span></TableCell>
              <TableCell align="left"><span className="app-table__header-cell">Date Time Ended</span></TableCell>
              <TableCell align="left"><span className="app-table__header-cell">Contractor Name</span></TableCell>
              <TableCell align="left"><span className="app-table__header-cell">Company</span></TableCell>
              <TableCell align="left"><span className="app-table__header-cell">Phone</span></TableCell>
              <TableCell align="center"><span className="app-table__header-cell">Status</span></TableCell>
              <TableCell align="center"></TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            <Suspense fallback={<Loader />}>
              <JobRowCreator
                jobData={data}
                showCompleted={displayCheckboxCompleted}
                showInCompleted={displayCheckboxInCompleted}
                descending={sortDescending}
                searchTerm={searchTerm}
              />
            </Suspense>
          </TableBody>
        </Table>
      </TableContainer>
    </>
  )
};


export default JobTable;
