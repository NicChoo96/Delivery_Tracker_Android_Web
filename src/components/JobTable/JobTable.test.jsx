import React from 'react';
import ReactDOM from 'react-dom';
import '@testing-library/jest-dom/extend-expect';

import { cleanup, wait } from '@testing-library/react';
import renderer from 'react-test-renderer';

import JobTable from './JobTable';

const mockData = [{ "name": "MalcomNew", "phoneNo": "69696969", "jobID": "-MJlA929nYBnioQAFsKp", "company": "MalcomCompany", "dateTimeEnd": 1602847819736, "dateTimeStart": 1602847809736, "locations": [] }, { "name": "MalcomNew", "phoneNo": "69696969", "jobID": "-MJlBeqaRX_7vLPeBU1m", "company": "MalcomCompany", "dateTimeEnd": 1602848216244, "dateTimeStart": 1602848206244, "locations": [] }, { "name": "MalcomNew", "phoneNo": "69696969", "jobID": "-MJlBrCGts_sRNdGfsZY", "company": "MalcomCompany", "dateTimeEnd": 1602848263811, "dateTimeStart": 1602848253811, "locations": [] }, { "name": "MalcomNew", "phoneNo": "69696969", "jobID": "-MJlLvAViWToCDvNQiUh", "company": "MalcomCompany", "dateTimeEnd": 1602850904558, "dateTimeStart": 1602850894558, "locations": [{ "dateTimeStamp": 1602851280074, "lat": 1.3, "lon": 1.2 }, { "dateTimeStamp": 1602851285192, "lat": 1.3, "lon": 1.2 }, { "dateTimeStamp": 1602851285965, "lat": 1.3, "lon": 1.2 }] }, { "name": "MalcomNew", "phoneNo": "69696969", "jobID": "-MJlO6j_fdnc4Bg5AXZL", "company": "MalcomCompany", "dateTimeEnd": 1602851480308, "dateTimeStart": 1602851470308, "locations": [{ "dateTimeStamp": 1602851470308, "lat": 1.69, "lon": 1.69 }, { "dateTimeStamp": 1602851470308, "lat": 1.69, "lon": 1.69 }] }, { "name": "Elon Musk", "phoneNo": "95123456", "jobID": "-MJqbgKIt7NWc7tWcZmu", "company": "MacroHard1", "dateTimeEnd": 0, "dateTimeStart": 1602939177220, "locations": [{ "dateTime": 1602939179290, "lat": 1.3482950000002, "lon": 103.683134 }, { "dateTime": 1602939181159, "lat": 2.3482950000002, "lon": 103.683134 }, { "dateTime": 1602939182328, "lat": 3.3482950000002, "lon": 103.683134 }, { "dateTime": 1602939183204, "lat": 4.3482950000002, "lon": 103.683134 }, { "dateTime": 1602939184106, "lat": 5.3482950000002, "lon": 103.683134 }] }, { "name": "Elon Musk", "phoneNo": "95123456", "jobID": "-MJqEZOEODr3OLokpCAw", "company": "MacroHard1", "dateTimeEnd": 0, "dateTimeStart": 1602932853205, "locations": [] }, { "name": "Elon Musk", "phoneNo": "95123456", "jobID": "-MJkxYjAhOAi8Rwol9H_", "company": "MacroHard1", "dateTimeEnd": 0, "dateTimeStart": 1602844244871, "locations": [] }, { "name": "Elon Musk", "phoneNo": "95123456", "jobID": "-MJm2HHKueP_7PEt5imA", "company": "MacroHard1", "dateTimeEnd": 0, "dateTimeStart": 1602862523652, "locations": [] }]

afterEach(cleanup);

describe('JobTable', () => {
  test('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<JobTable />, div);
  });

  test('comparing generated rows with file count', async () => {
    const component = (
      <JobTable data-testid="jobtable__testcount" data={mockData} />
    );
    await wait();
    const jowRowCount = component.props.data.length;
    const fileCount = mockData.length;
    expect(jowRowCount).toEqual(fileCount);
  });

  test('matches snapshot', () => {
    const tree = renderer.create(<JobTable />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
