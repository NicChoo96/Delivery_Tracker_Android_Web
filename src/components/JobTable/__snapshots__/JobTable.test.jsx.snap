// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobTable matches snapshot 1`] = `
Array [
  <div
    className="job-table__data-control-group"
  >
    <div
      className="MuiFormGroup-root MuiFormGroup-row"
    >
      <label
        className="MuiFormControlLabel-root"
      >
        <span
          aria-disabled={false}
          className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-checked-2 Mui-checked MuiIconButton-colorPrimary"
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label"
          >
            <input
              checked={true}
              className="PrivateSwitchBase-input-4"
              data-indeterminate={false}
              name="Show Completed"
              onChange={[Function]}
              type="checkbox"
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
          </span>
        </span>
        <span
          className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          In Progress
        </span>
      </label>
      <label
        className="MuiFormControlLabel-root"
      >
        <span
          aria-disabled={false}
          className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-checked-2 Mui-checked MuiIconButton-colorPrimary"
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label"
          >
            <input
              checked={true}
              className="PrivateSwitchBase-input-4"
              data-indeterminate={false}
              name="Show Incompleted"
              onChange={[Function]}
              type="checkbox"
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
          </span>
        </span>
        <span
          className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Completed
        </span>
      </label>
      <label
        className="MuiFormControlLabel-root"
      >
        <span
          className="MuiSwitch-root"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-checked-2 Mui-checked"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label"
            >
              <input
                checked={true}
                className="PrivateSwitchBase-input-4 MuiSwitch-input"
                name="Show Descending"
                onChange={[Function]}
                type="checkbox"
              />
              <span
                className="MuiSwitch-thumb"
              />
            </span>
          </span>
          <span
            className="MuiSwitch-track"
          />
        </span>
        <span
          className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Date Time Descending
        </span>
      </label>
    </div>
    <div
      className="MuiFormGroup-root MuiFormGroup-row"
    >
      <label
        className="MuiFormControlLabel-root"
      >
        <div
          className="MuiFormControl-root MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
            data-shrink={false}
          >
            Search for Contractor
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
              disabled={false}
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="PrivateNotchedOutline-legendLabelled-7"
              >
                <span>
                  Search for Contractor
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <span
          className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        />
      </label>
    </div>
  </div>,
  <div
    className="MuiPaper-root MuiTableContainer-root app-table MuiPaper-elevation1 MuiPaper-rounded"
  >
    <table
      aria-label="a dense table"
      className="MuiTable-root"
      role={null}
    >
      <thead
        className="MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeSmall"
            scope="col"
            style={
              Object {
                "width": "8em",
              }
            }
          >
            <span
              className="app-table__header-cell"
            >
              Job ID
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeSmall"
            scope="col"
          >
            <span
              className="app-table__header-cell"
            >
              Location
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeSmall"
            scope="col"
          >
            <span
              className="app-table__header-cell"
            >
              Date Time Started
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeSmall"
            scope="col"
          >
            <span
              className="app-table__header-cell"
            >
              Date Time Ended
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeSmall"
            scope="col"
          >
            <span
              className="app-table__header-cell"
            >
              Contractor Name
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeSmall"
            scope="col"
          >
            <span
              className="app-table__header-cell"
            >
              Company
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeSmall"
            scope="col"
          >
            <span
              className="app-table__header-cell"
            >
              Phone
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeSmall"
            scope="col"
          >
            <span
              className="app-table__header-cell"
            >
              Status
            </span>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeSmall"
            scope="col"
          />
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root"
          role={null}
        >
          <td
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeSmall"
            colSpan={8}
          >
            <span
              className="app__no-data-message"
            >
              No data available.
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
]
`;
